apps/hash-ai-worker-ts/src/activities/shared/get-flow-context.ts

```diff
@@ -0,0 +1,108 @@
+import { createGraphClient } from "@local/hash-backend-utils/create-graph-client";
+import { createTemporalClient } from "@local/hash-backend-utils/temporal";
+import { parseHistoryItemPayload } from "@local/hash-backend-utils/temporal/parse-history-item-payload";
+import type { GraphApi } from "@local/hash-graph-client";
+import { getRequiredEnv } from "@local/hash-isomorphic-utils/environment";
+import type { RunFlowWorkflowParams } from "@local/hash-isomorphic-utils/flows/temporal-types";
+import type {
+  AccountId,
+  EntityId,
+  EntityUuid,
+  OwnedById,
+} from "@local/hash-subgraph";
+import { entityIdFromComponents } from "@local/hash-subgraph";
+import { Context } from "@temporalio/activity";
+
+import { logToConsole } from "../../shared/logger";
+
+const temporalClient = await createTemporalClient();
+
+const getWorkflowExecutionStartedEventAttributesOfWorkflow = async (params: {
+  workflowId: string;
+}) => {
+  const { workflowId } = params;
+
+  const handle = temporalClient.workflow.getHandle(workflowId);
+
+  const { events } = await handle.fetchHistory();
+
+  if (!events) {
+    throw new Error(`No events found for workflowId ${workflowId}`);
+  }
+
+  const workflowExecutionStartedEventAttributes =
+    events.find((event) => event.workflowExecutionStartedEventAttributes)
+      ?.workflowExecutionStartedEventAttributes ?? undefined;
+
+  return { workflowExecutionStartedEventAttributes };
+};
+
+type FlowContext = {
+  flowEntityId: EntityId;
+  graphApiClient: GraphApi;
+  userAuthentication: { actorId: AccountId };
+  webId: OwnedById;
+};
+
+const graphApiClient = createGraphClient(logToConsole, {
+  host: getRequiredEnv("HASH_GRAPH_API_HOST"),
+  port: parseInt(getRequiredEnv("HASH_GRAPH_API_PORT"), 10),
+});
+
+/**
+ * Get the context of the flow that is currently being executed
+ * from a temporal activity.
+ *
+ * This method must be called from a temporal activity that is
+ * called within the `runFlow` temporal workflow.
+ */
+export const getFlowContext = async (): Promise<FlowContext> => {
+  const activityContext = Context.current();
+
+  const { workflowId } = activityContext.info.workflowExecution;
+
+  /**
+   * @todo: consider caching the result of this, so that multiple calls
+   * to `getFlowContext` within the same flow do not result in individual
+   * calls to the Temporal API.
+   */
+  const { workflowExecutionStartedEventAttributes } =
+    await getWorkflowExecutionStartedEventAttributesOfWorkflow({
+      workflowId,
+    });
```

Comment:
Given how frequently we will need to call this function I would add a cache now. Could use [cache-manager](https://www.npmjs.com/package/cache-manager). I've used [node-cache](https://www.npmjs.com/package/node-cache) but it's no longer maintained.

Reply:
Fair enough, have added the cache in https://github.com/hashintel/hash/pull/4379/commits/5de27c1ac9faca589e36f74e10cd04aabf5cbf3f. We'll only be storing the parts of the params we are actually returning in `getFlowContext`, and I've set a maximum of 100 items (100 concurrent "flows") and a time to live of 10 minutes (accounting for longer running flows). The data is fairly minimal so I'm not concerned about the cache size, but if we add things down the line those parameters might need to be adjusted.